rcode-reader/src/main/java/eu/rssw/pct/PLReader.java

```diff
@@ -0,0 +1,174 @@
+
+  public PLReader(File file) {
+    String name = file.getPath();
+    SecurityManager sm = System.getSecurityManager();
+    if (sm != null) {
+      sm.checkRead(name);
+    }
+
+    this.pl = file;
+  }
+
+  /**
+   * Returns entries contained in this procedure library
+   * 
+   * @throws RuntimeException If file is not a valid procedure library
+   */
+  public List<FileEntry> getFileList() {
+    if (this.files == null)
+      readFileList();
+    return files;
+  }
+
+  public FileEntry getEntry(String name) {
+    for (FileEntry entry : getFileList()) {
+      if (entry.getFileName().equals(name))
+        return entry;
+    }
+
+    return null;
+  }
+
+  private void readFileList() {
+    try (RandomAccessFile raf = new RandomAccessFile(pl, "r")) {
+      FileChannel fc = raf.getChannel();
+      ByteBuffer magic = ByteBuffer.allocate(2);
+      fc.read(magic);
+      if ((magic.getShort(0) & 0xffff) != MAGIC_V11)
+        throw new RuntimeException("Not a valid PL file");
+
+      Charset charset = getCharset(fc);
+      int offset = getTOCOffset(fc);
+      files = new ArrayList<>();
+      FileEntry fe = null;
+      while ((fe = readEntry(fc, offset, charset)) != null) {
+        if (fe.isValid())
+          files.add(fe);
+        offset += fe.getTocSize();
+      }
+    } catch (IOException caught) {
+      throw new RuntimeException(caught);
+    }
+  }
+
+  public InputStream getInputStream(FileEntry fe) throws IOException {
+    ByteBuffer bb = null;
+    try (RandomAccessFile raf = new RandomAccessFile(pl, "r")) {
+      FileChannel fc = raf.getChannel();
+      bb = ByteBuffer.allocate(fe.getSize());
+      fc.read(bb, fe.getOffset());
+    }
+
+    return new ByteArrayInputStream(bb.array());
+  }
+
+  private Charset getCharset(FileChannel fc) throws IOException {
+    ByteBuffer bEncoding = ByteBuffer.allocate(ENCODING_SIZE);
+    if (fc.read(bEncoding, ENCODING_OFFSET) != ENCODING_SIZE)
+      throw new RuntimeException("Invalid PL file");
+    bEncoding.position(0);
+    StringBuilder sbEncoding = new StringBuilder();
+    int zz = 0;
+    while ((zz < 20) && (bEncoding.get(zz) != 0)) {
+      sbEncoding.append((char) bEncoding.get(zz++));
+    }
+    try {
+      return Charset.forName(sbEncoding.toString());
+    } catch (IllegalArgumentException iae) {
+      return Charset.forName("US-ASCII");
+    }
+  }
+
+  private int getTOCOffset(FileChannel fc) throws IOException {
+    ByteBuffer bTOC = ByteBuffer.allocate(4);
+    if (fc.read(bTOC, FILE_LIST_OFFSET_V11) != 4)
+      throw new RuntimeException("Invalid PL file");
+    return bTOC.getInt(0);
+  }
+
+  private FileEntry readEntry(FileChannel fc, int offset, Charset charset) throws IOException {
+    ByteBuffer b1 = ByteBuffer.allocate(1);
+    fc.read(b1, offset);
+
+    if (b1.get(0) == (byte) 0xFE) {
+      boolean stop = false;
+      int zz = 0;
+      while (!stop) {
+        b1.position(0);
+        int kk = fc.read(b1, offset + ++zz);
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Cast one of the operands of this addition operation to a "long". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://sonar.riverside-software.fr/coding_rules#rule_key=squid%3AS2184)